internal/hcsoci/resources.go

```diff
@@ -23,158 +20,81 @@ func (r *Resources) NetNS() string {
+	Release(context.Context) error
 }
 
-// TODO: Method on the resources?
+// ReleaseResources releases/frees all of the resources associated with a container. This includes
+// Plan9 shares, vsmb mounts, pipe mounts, network endpoints, scsi mounts, vpci devices and layers.
 func ReleaseResources(ctx context.Context, r *Resources, vm *uvm.UtilityVM, all bool) error {
-	if vm != nil && r.addedNetNSToVM {
-		if err := vm.RemoveNetNS(ctx, r.netNS); err != nil {
-			log.G(ctx).Warn(err)
-		}
-		r.addedNetNSToVM = false
-	}
-
-	if r.createdNetNS {
-		for len(r.networkEndpoints) != 0 {
-			endpoint := r.networkEndpoints[len(r.networkEndpoints)-1]
-			err := hns.RemoveNamespaceEndpoint(r.netNS, endpoint)
-			if err != nil {
-				if !os.IsNotExist(err) {
-					return err
-				}
-				log.G(ctx).WithFields(logrus.Fields{
-					"endpointID": endpoint,
-					"netID":      r.NetNS(),
-				}).Warn("removing endpoint from namespace: does not exist")
-			}
-			r.networkEndpoints = r.networkEndpoints[:len(r.networkEndpoints)-1]
-		}
-		r.networkEndpoints = nil
-		err := hns.RemoveNamespace(r.netNS)
-		if err != nil && !os.IsNotExist(err) {
-			return err
-		}
-		r.createdNetNS = false
-	}
-
-	if vm != nil && all {
-		for len(r.vsmbMounts) != 0 {
-			mount := r.vsmbMounts[len(r.vsmbMounts)-1]
-			if err := vm.RemoveVSMB(ctx, mount); err != nil {
-				return err
-			}
-			r.vsmbMounts = r.vsmbMounts[:len(r.vsmbMounts)-1]
-		}
-
-		for len(r.pipeMounts) != 0 {
-			mount := r.pipeMounts[len(r.pipeMounts)-1]
-			if err := vm.RemovePipe(ctx, mount); err != nil {
-				return err
+	if vm != nil {
+		if r.addedNetNSToVM {
+			if err := vm.RemoveNetNS(ctx, r.netNS); err != nil {
+				log.G(ctx).Warn(err)
 			}
-			r.pipeMounts = r.pipeMounts[:len(r.pipeMounts)-1]
+			r.addedNetNSToVM = false
 		}
 
-		for len(r.plan9Mounts) != 0 {
-			mount := r.plan9Mounts[len(r.plan9Mounts)-1]
-			if err := vm.RemovePlan9(ctx, mount); err != nil {
-				return err
+		// vpci cleanup
+		for _, name := range r.vpciDevices {
+			if err := vm.RemoveDevice(ctx, name); err != nil {
+				log.G(ctx).WithError(err).Warn("failed to remove VPCI device")
 			}
-			r.plan9Mounts = r.plan9Mounts[:len(r.plan9Mounts)-1]
 		}
 
-		for _, sm := range r.scsiMounts {
-			if err := vm.RemoveSCSI(ctx, sm.path); err != nil {
-				return err
-			}
-			if sm.autoManage {
-				if err := os.Remove(sm.path); err != nil {
-					log.G(ctx).WithError(err).Warnf("failed to remove automanage-virtual-disk at: %q", sm.path)
+		if all {
+			// For now the resources that implement the ReleaseCloser interface are vsmb shares,
+			// scsi mounts, pipe mounts, and plan9 mounts. The goal is to move over networking,
+			// vpci devices, and layers to this so they all follow the same format.
+			for len(r.resources) != 0 {
+				resource := r.resources[len(r.resources)-1]
+				if resource != nil {
+					if err := resource.Release(ctx); err != nil {
+						return err
+					}
 				}
+				r.resources = r.resources[:len(r.resources)-1]
```

Comment:
For this, could we iterate over the resources like 
` for _, resource := range r.resources` and then just nil the r.resources slice afterwards? That avoids the need to reslice the slice every time we release something.